### PR TITLE
UI Docs update

### PR DESCRIPTION
**Pull request by aspiringLich**
This is to go along with the pr space-wizards/space-station-14#29903.

In this PR I:

*   Mark the previous UI guide as outdated (don't delete for now, because downstream forks or something)
*   Move the "documentation" for the `Robust.Client` controls into their own doc (for some poor sod who wants to add more) (that poor sod may be me)
*   "UI and You" may be the greatest piece of literature to date, and covers the new styling system very well, and makes a half-hearted attempt at explaining C# UI code (someone who knows this better please fix)
*   "UI Cookbook" so people know things exist, and then they like, use them, I guess.

### CODE REVIEW

File: src/en/ss14-by-example/ui-and-you.md [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+        .Font(sheet.BaseFont.GetFont(8))
+];
+```
+
+### Death to Hardcoding!
+
+You may have noticed before that there wasn't much hardcoding in the rule definitions.
+This is because most style rules are used in multiple stylesheets, and they have
+to be generic between them. There are a couple different utilities used to reduce
+hardcoding.
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:183]: This paragraph feels a bit awkward to me. I think I'd reword this focusing more on the purpose behind avoiding hardcoding (increasing reusability of common patterns).

File: src/en/ss14-by-example/ui-and-you.md [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+
+#### `ColorPalette`
+
+There is actually a pretty robust (haha) color palette system to hopefully make
+hardcoding colors unnecessary. I created a
+[codepen](https://codepen.io/aspiringLich/pen/VwOXdjd?editors=1000)
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:189]: Linking to an external website like this feels like a recipe for a future dead or outdated link.  I'm not sure what the best alternative is, though.

File: src/en/ss14-by-example/ui-and-you.md [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+to help visualize the colors. The stylesheets then use these palettes to set the following
+common palettes for sheetlets to reference:
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:191]: This last sentence is weird. The stylesheets use palettes to set palettes? What does it means for sheetlets to reference something? I think it should be more direct about what actual effect palettes have.

File: src/en/ss14-by-example/ui-and-you.md [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+
+-   `PrimaryPalette`: Used for foreground elements
+-   `SecondaryPalette`: Used for Background elements
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:194]:
```suggestion
-   `SecondaryPalette`: Used for background elements
```

File: src/en/ss14-by-example/ui-and-you.md [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+
+In C#, you access the colors by accessing the
+properties on the `ColorPalette` class. From brightest to darkest, the properties
+are (as of writing) arranged as so:
+
+-   `+0`: `Text` `Base`
+-   `-1`: `TextDark`, `Element`
+-   `-2`: `BackgroundLight`, `PressedElement`
+-   `-3`: `Background`
+-   `-4`: `BackgroundDark`, `DisabledElement`
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:207]: What are the numbers here?

File: src/en/ss14-by-example/ui-and-you.md [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+
+Although it would be simpler to have the palettes be arrays of colors, I noticed
+that colors were used in basically three ways: Text, foreground elements, and
+background elements. So I represented that in the palette itself! It just makes
+the code a bit more readable.
+
+All palettes are defined in `Palettes.cs` as static properties. You can read
+them from anywhere! Neat!
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:215]: These last two paragraphs could be re-written. I think it'd be better to just describe the why of why this approach was taken.

File: src/en/ss14-by-example/ui-and-you.md [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+
+#### `ISheetletConfig`
+
+`ISheetletConfig` is intended to cut down on repeated code by providing shared
+functionality and definitions between the stylesheets.
+Any `Sheetlet` that requires the values in some instance of
+`ISheetletConfig` should have a generic type constraint that requires the
+`ISheetletConfig` interface.
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:223]: This needs a better example. What kind of repeated code? What functionality is being shared?

File: src/en/ss14-by-example/ui-and-you.md [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+```
+
+#### Resource Access
+
+You should access resources differently in sheetlets. Each stylesheet provides a
+list of resource roots to look in, when requesting a resource of a certain type.
+For example, the root for `NanotrasenStylesheet` for `TextureResource` is
+`/Textures/Interface/Nano`. Any texture requested with `GetTexture` will will
+append the provided relative path to the root, and return the texture if it exists.
+
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:246]:
```suggestion
Resources are accessed different in sheetlets than other parts of the codebase.
Each stylesheet provides a list of directories (roots) to use when a resource
is requested. For example, the `TextureResource` root in `NanotrasenStylesheet`
is `/Textures/Interface/Nano`. This means that any texture requested with
`GetTexture` will be searched for relative to this directory.
```

File: src/en/ss14-by-example/ui-and-you.md [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+This system is built to work with any resource type, though currently only textures
+are used in the sheetlets.
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:248]: This paragraph isn't really necessary. The section is already about accessing resources in general.

File: src/en/ss14-by-example/ui-and-you.md [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+
+### Generic Sheetlets
+
+As stated earlier, generic sheetlets are used for generic UI elements that are
+used in many different UIs.
+
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:254]:
```suggestion
Generic sheetlets are used for common UI elements that are used throughout many
different UIs.
```

File: src/en/ss14-by-example/ui-and-you.md [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+-   You should always select elements with `.Class` and not `.Identifier`.
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:255]: These bullet points need an introduction (here are some best practicies blah blah).

File: src/en/ss14-by-example/ui-and-you.md [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+-   When accessing resources, use the `GetTextureOr` method to get the texture and
+    provide a fallback root to use if the texture is not found within the stylesheets roots.
+-   Avoid manual hardcoding of classes. When referencing classes,
+    please only use classes defined on the element being styled (in `StyleClass*` properties)
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:259]: Using 'please' in technical documentation is very weird to me. In my opinion it should pretty much always be a 'you should' instead.

File: src/en/ss14-by-example/ui-and-you.md [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+    or define your own in `StyleClass.cs`.
+-   Avoid manual hardcoding of colors. Use the palettes provided by the stylesheet
+    to set the colors of elements. There are very few cases where you will actually
+    need to manually hardcode a color in a generic sheetlet.
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:263]:
```suggestion
-   Use the palettes provided by the set stylesheet instead of specifying 
    colors by hand. There are very few cases where you will actually need
    to manually hardcode a color in a generic sheetlet.
```

File: src/en/ss14-by-example/ui-and-you.md [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+-   If you need to access a resource that is not provided already, please add the path
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:264]: Ditto the note on 'please' (and also anywhere else); I see one other usage earlier that I didn't mention too.

File: src/en/ss14-by-example/ui-and-you.md [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+```
+
+</details>
+
+### Specific Sheetlets
+
+As stated earlier, specific sheetlets are used for UI elements that are specific
+to a single UI. These sheetlets are located with the `*.xaml` file they are
+associated with.
+
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:319]:
```suggestion
Specific sheetlets are used in conjunction with UI elements that are used only
a few times, most often all in a single UI. These sheetlets are located in the
same directory as the `*.xaml` file they are associated with.
```

File: src/en/ss14-by-example/ui-and-you.md [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+-   You should prefer to select elements with `.Identifier` and not `.Class`.
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:320]: This bullet list needs an intro stating what these points are about.

File: src/en/ss14-by-example/ui-and-you.md [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+// which stylesheet is this sheetlet for
+public sealed class PaperSheetlet : Sheetlet<NanotrasenStylesheet>
+{
+    public override StyleRule[] GetRules(NanotrasenStylesheet sheet, object config)
+    {
+        // define any IConfigs you need here
+        IWindowConfig windowCfg = sheet;
+
+        // get any textures / construct any complicated resources here
+        var paperBackground = ResCache.GetTexture("/Textures/Interface/Paper/paper_background_default.svg.96dpi.png")
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:356]: This filename should be a static readonly respath. I don't think we validate those, but we probably will some day.

File: src/en/ss14-by-example/ui-and-you.md [hunk 18]
```diff
@@ -0,0 +1,474 @@
...
+        ];
+    }
+}
+```
+
+</details>
+
+### Making your own `Stylesheet`
+
+Creating a new stylesheet is pretty simple, so I won't go into it too much.
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:378]: It's simple because you wrote the whole dang thing :P

Something like making a whole stylesheet is absolutely not simple to anyone else. Since this section doesn't really say anything at all about how to make a stylesheet, I think it's fine to just remove it, or mark that the section needs expanding (which is super fine to do!).

Also evil first person in technical documentation!!

File: src/en/ss14-by-example/ui-and-you.md [hunk 19]
```diff
@@ -0,0 +1,474 @@
...
+One thing to keep in mind is how colors are chosen.
+
+The colors are generated using the [OKLAB color space](https://bottosson.github.io/posts/oklab/),
+which is better because something something human eyes something. When you choose
+new colors for your stylesheet, it may be helpful to use an [OKLCH Color Picker](https://oklch.com)
+and modify an existing color.
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:384]: Pedantry time: colors aren't really generated using a color space—in fact, a color on its own doesn't "have" a color space. A color space is really just a way of organizing colors (usually implicitly paired with a color model which is how we represent those colors—that's the difference between Oklab and OKLCH).

It's also fine to just hand off the explanation of why Oklab is cool to another website, but the gist of it is that it lets us interpolate between certain parameters in a way that visually preserves the unchanged variables. A transition between two colors of equal lightness will have perceptually uniform lightness along the entire transition. This ends up being very helpful for anything that ends up being 'what people see', as opposed to other color spaces and models that are more useful for (some) computations.

File: src/en/ss14-by-example/ui-and-you.md [hunk 20]
```diff
@@ -0,0 +1,474 @@
...
+
+The best way to learn how to write UI code is to look at existing code. Some UIs
+definitely do some terrible things you should never replicate, but there are mountains
+of terrible code in SS14, so this is not abnormal. I cannot teach familiarity with
+the internals of this game, but I can give a  general overview.
+
+Code you could reference:
+-    Robotics Console
+-    Cryo Pod
+-    Reagent Dispenser
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:400]: BatteryMenu is probably the best example right now, expect for the fact that it is still BUI state based.

File: src/en/ss14-by-example/ui-and-you.md [hunk 21]
```diff
@@ -0,0 +1,474 @@
...
+One thing to keep in mind is how colors are chosen.
+
+The colors are generated using the [OKLAB color space](https://bottosson.github.io/posts/oklab/),
+which is better because something something human eyes something. When you choose
+new colors for your stylesheet, it may be helpful to use an [OKLCH Color Picker](https://oklch.com)
+and modify an existing color.
+
+## Writing C# for UI
+
+> **TODO:** I don't feel confident enough in my knowledge to describe in detail
```
Comment (perryprog) [inline src/en/ss14-by-example/ui-and-you.md:388]: I think talking about actual UI implementation can be done in a different page, and you could definitely get away with not having to write about it in this PR if you want :P

**Review by perryprog — COMMENTED** [review]
Uhhhhh here's a few nitpicks and comments. Feel free to disregard anything you like. Some of the suggested changes might have bad line selections but it should be clear what the intent was.

Main general comment is that technical documentation shouldn't use the first person, and it should _definitely_ avoid using subjective descriptions of code complexity or implementation difficulty.

File: src/en/ss14-by-example/ui-and-you/ui-cookbook.md [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+arrays in XAML are easy-peasy to define!
+
+```xml
+<!-- stupid and bad, why would you ever think this would work?? -->
+<Control>
+    <Button StyleClasses="ButtonSmall negative" Text="delete" />
+</Control>
+
+<!-- the correct solution (obviously) -->
+<Control xmlns:sys="clr-namespace:System;assembly=System.Runtime">
```
[the remark this answers is not in the document]
Comment (perryprog) [reply src/en/ss14-by-example/ui-and-you/ui-cookbook.md:37]: Oh, could also mention that this is something you can do on the C# side as well which can be easier.

File: src/SUMMARY.md [hunk 23]
```diff
@@ -51,9 +51,10 @@ SS14 By Example
   - [Porting Appearance Visualizers](en/ss14-by-example/making-a-sprite-dynamic/porting-appearance-visualizers.md)
 - [Basic Networking and You](en/ss14-by-example/basic-networking-and-you.md)
 - [Fluent and Localization](en/ss14-by-example/fluent-and-localization.md)
+- [UI and You](en/ss14-by-example/ui-and-you.md)
```
Comment (Fildrance) [inline src/SUMMARY.md]: i know that its funny-ish but the joke was previously reserved by networking section, so probably better call it 'UI basics' or something

File: src/en/ss14-by-example/ui-and-you.md [hunk 24]
```diff
@@ -0,0 +1,474 @@
...
+```admonish note
+Previously every single style rule was in one giant list: `StyleNano.cs`, a 1600
+line pit of despair where dreams went to die.
+```
+
+There are, primarily, two types of Sheetlets:
+
+-   **Generic Sheetlets**: These go in `Content.Client/Stylesheets/Sheetlets`.
+    These stylesheets concern generic UI elements used in many different UIs, and
+    should be written generically to work with any stylesheet.
```
Comment (Fildrance) [inline src/en/ss14-by-example/ui-and-you.md]: you mean 'should be written to work with any OTHER stylesheet'?

File: src/en/ss14-by-example/ui-and-you.md [hunk 25]
```diff
@@ -0,0 +1,474 @@
...
+to help visualize the colors. The stylesheets then use these palettes to set the following
+common palettes for sheetlets to reference:
+
+-   `PrimaryPalette`: Used for foreground elements
+-   `SecondaryPalette`: Used for Background elements
+-   `PositivePalette`: A traditionally green palette used to represent success / good / full
+-   `NegativePalette`: A traditionally red palette used to represent errors / bad / empty
+-   `HighlightPalette`: Used to highlight headings or important elements
+
+In C#, you access the colors by accessing the
```
Comment (Fildrance) [inline src/en/ss14-by-example/ui-and-you.md]: This palette section sadly seems a bit incoherent and scuffed - maybe best way is to show some examples?

File: src/en/ss14-by-example/ui-and-you.md (hunk 18 above)
Comment (Fildrance) [reply src/en/ss14-by-example/ui-and-you.md]: On top of that i have to remark that it is critical to mention when to use stylesheet and when - not. Having several examples will be critical. This stuff is VERY not trivial for newcomers, and even for seasoned veterans - stylesheets does not always sound very alluring, compared to just setting couple of values in XAML or in control c-tor. And sometimes it can work.

File: src/en/ss14-by-example/ui-and-you.md [hunk 26]
```diff
@@ -0,0 +1,474 @@
...
+does, its as simple as looking at the field's usages, and reading the style rule definition.
+This is where it may be helpful to use the Rider IDE, as it has a fantastic implementation of this feature,
+but its probably possible in other IDEs as well.
+
+```admonish tip
+In general, if you are doing UI dev, I would recommend using the Rider IDE. It
+eats up quite a bit of RAM, but it provides autocomplete in XAML files, a lot
+of really nice auto-refactoring and searching features, and very decent git
+integration. Give it a shot!
+```
```
[the remark this answers is not in the document]
Comment (aspiringLich) [reply src/en/ss14-by-example/ui-and-you.md]: I'm pretty sure rider is better? I'll just make a note about VSCode in case im dumb

File: src/en/ss14-by-example/ui-and-you.md [hunk 27]
```diff
@@ -0,0 +1,474 @@
...
+or resources that could be commonly used.
+
+### All hail the mighty `Sheetlet`
+
+It's important to understand that basically, a stylesheet is a massive list of
+every single style rule. Instead of manually
+making a giant list of style rules (because that would be ridiculous... haha... ha....),
+the responsibility of chipping into this list is distributed between many Sheetlets.
+Each Sheetlet returns a small chunk of style rules, which is agglomerated into the
+final list at the end.
```
[the remark this answers is not in the document]
Comment (aspiringLich) [reply src/en/ss14-by-example/ui-and-you.md]: I'm going to be so real I'm not sure. I just elaborated a bit on the specificity of style rules a bit. That's the only surefire way to get style rules to behave.

[the remark this answers is not in the document]
Comment (aspiringLich) [reply src/en/ss14-by-example/ui-and-you.md]: Also I'm leaking the details so people fully understand why stylenano existed

File: src/SUMMARY.md (hunk 23 above)
Comment (aspiringLich) [reply src/SUMMARY.md]: That's why I picked it for like, symmetry! Or something. I think UI survival guide was a good name but that guide should stick around for a little while for transitionary purposes.

File: src/en/ss14-by-example/ui-and-you.md [hunk 28]
```diff
@@ -0,0 +1,474 @@
...
+
+-   `Type`: The type of element this rule affects. Anything inheriting from this
+    type will be affected by this rule. This is similar to the element selector
+    in css.
+-   `StyleClasses`: The classes that the element must have to be affected by this
+    rule. The element must have all of these classes to be affected by this rule.
+-   `StyleIdentifier`: The identifier of the element. This is a unique identifier
+    that can be used to target a specific element. This should be used sparingly,
+    when there is only one instance of the element that needs to be styled in a
+    highly specific manner. There may only be one of these specified.
```
[the remark this answers is not in the document]
Comment (aspiringLich) [reply src/en/ss14-by-example/ui-and-you.md]: I tried my best

File: src/en/ss14-by-example/ui-and-you.md (hunk 17 above)
Comment (aspiringLich) [reply src/en/ss14-by-example/ui-and-you.md]: Uh oh I don't think I do that in the like... actual PR. I'll... leave it for now.

File: src/en/ss14-by-example/ui-and-you.md [hunk 29]
```diff
@@ -0,0 +1,474 @@
...
+> Prefer to use the overloads that take in EntityUids instead of ICommonSession, this will make it easier to code NPCs who can interact with UIs in future.
+>
+> >
+>
+> -   There's a helper under this.CreateWindow<TWindow>() for BUIs that handles disposing + opening + close subscription for you.
+> -   There's a method OnProtoReload that gets called on BUIs so you can override and handle it without having to manually subscribe on another system.
+> -   I have added prototype reload support to some stuff.
+> -   I have cleaned up a lot of BUI code. Windows now just raise events and the BUI itself handles message sending.
+>
+> Some notes for future:
```
[the remark this answers is not in the document]
Comment (aspiringLich) [reply src/en/ss14-by-example/ui-and-you.md]: ditto

File: src/en/ss14-by-example/ui-and-you.md (hunk 2 above)
Comment (aspiringLich) [reply src/en/ss14-by-example/ui-and-you.md]: Replaced with an image.

File: src/en/robust-toolbox/user-interface/containers.md [hunk 30]
```diff
@@ -0,0 +1,85 @@
+# Container Controls
+
+Container is type of control that dictates the layout of
+other controls. Containers logic affect directly only direct child controls.
+Containers usually arrange child controls every frame by collecting desired size
+of controls (call to `Measure` method on children) and then providing them with
+size that they can afford (call to `Arrange` method on children), and placing
```
Comment (juliangiebel) [inline src/en/robust-toolbox/user-interface/containers.md]:
```suggestion
the size that they can afford by calling the `Arrange` method on them.
```

File: src/en/robust-toolbox/user-interface/containers.md [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+them (`ArrangeOverride`) at needed coordinates after all calculations were
```
Comment (juliangiebel) [inline src/en/robust-toolbox/user-interface/containers.md]:
```suggestion
After calculation the desired and afforded sizes of its children the container can override the child components coordinates if needed.
```

File: src/en/robust-toolbox/user-interface.md [hunk 32]
```diff
@@ -1,5 +1,9 @@
 # User Interface
 
+```admonish warning
+The information on this page may be outdated, and may not accurately describe current conventions for upstream Space Station 14.
+```
+
```
Comment (juliangiebel) [inline src/en/robust-toolbox/user-interface.md]: Shouldn't this use `{{ #template ../templates/outdated.md }}`?

File: src/en/robust-toolbox/user-interface/containers.md [hunk 33]
```diff
@@ -0,0 +1,85 @@
...
+accounted.
```
Comment (juliangiebel) [inline src/en/robust-toolbox/user-interface/containers.md]:
```suggestion
other controls. The containers logic only affects direct child controls.
Containers usually arrange child controls every frame by collecting their desired size by calling the `Measure` method on them and then providing them with the size that they can afford by calling the `Arrange` method on them.
After calculation the desired and afforded sizes of its children the container can override the child components coordinates if needed.
```

**Review by juliangiebel — CHANGES_REQUESTED** [review]
Pretty good documentation and the new stylesheet and sheetlet system **really** shouldn't be sitting around undocumented.
The container documentation has a paragraph that is kinda hard to read. I suggested a way to untangle it a bit but feel free to write something yourself.
